sycl/source/handler.cpp

```diff
@@ -842,6 +857,280 @@ void handler::ext_oneapi_memset2d_impl(void *Dest, size_t DestPitch, int Value,
+          : (SrcImgDesc.height > 0 ? PI_MEM_TYPE_IMAGE2D : PI_MEM_TYPE_IMAGE1D);
+
+  sycl::detail::pi::PiMemImageFormat PiFormat;
+  PiFormat.image_channel_data_type =
+      sycl::_V1::detail::convertChannelType(SrcImgDesc.channel_type);
+  PiFormat.image_channel_order =
+      sycl::_V1::detail::convertChannelOrder(SrcImgDesc.channel_order);
+
+  MImpl->MSrcOffset = {SrcOffset[0], SrcOffset[1], SrcOffset[2]};
+  MImpl->MDestOffset = {DestOffset[0], DestOffset[1], DestOffset[2]};
+  MImpl->MCopyExtent = {CopyExtent[0], CopyExtent[1], CopyExtent[2]};
+  MImpl->MHostExtent = {DestExtent[0], DestExtent[1], DestExtent[2]};
+  MImpl->MImageDesc = PiDesc;
+  MImpl->MImageFormat = PiFormat;
+  MImpl->MImageCopyFlags =
+      sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_DEVICE_TO_HOST;
+  setType(detail::CG::CopyImage);
+}
+
+void handler::ext_oneapi_copy(
+    void *Src, void *Dest,
+    const ext::oneapi::experimental::image_descriptor &Desc, size_t Pitch) {
+  MSrcPtr = Src;
+  MDstPtr = Dest;
+
+  sycl::detail::pi::PiMemImageDesc PiDesc = {};
+  PiDesc.image_width = Desc.width;
+  PiDesc.image_height = Desc.height;
+  PiDesc.image_depth = Desc.depth;
+  PiDesc.image_type = Desc.depth > 0 ? PI_MEM_TYPE_IMAGE3D
+                                     : (Desc.height > 0 ? PI_MEM_TYPE_IMAGE2D
+                                                        : PI_MEM_TYPE_IMAGE1D);
+
+  sycl::detail::pi::PiMemImageFormat PiFormat;
+  PiFormat.image_channel_data_type =
+      sycl::_V1::detail::convertChannelType(Desc.channel_type);
+  PiFormat.image_channel_order =
+      sycl::_V1::detail::convertChannelOrder(Desc.channel_order);
+
+  // Flags
+  sycl::detail::pi::PiImageCopyFlags PiCopyFlags;
+  sycl::usm::alloc DstPtrType = get_pointer_type(Dest, MQueue->get_context());
+  sycl::usm::alloc SrcPtrType = get_pointer_type(Src, MQueue->get_context());
+  if (DstPtrType == sycl::usm::alloc::device) {
+    // Dest is on device
+    if (SrcPtrType == sycl::usm::alloc::device) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_DEVICE_TO_DEVICE;
+    } else if (SrcPtrType == sycl::usm::alloc::host ||
+               SrcPtrType == sycl::usm::alloc::unknown) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_HOST_TO_DEVICE;
+    } else {
+      assert(false && "Unknown copy source location");
+    }
+  } else if (DstPtrType == sycl::usm::alloc::host ||
+             DstPtrType == sycl::usm::alloc::unknown) {
+    // Dest is on host
+    if (SrcPtrType == sycl::usm::alloc::device) {
+      PiCopyFlags =
+          sycl::detail::pi::PiImageCopyFlags::PI_IMAGE_COPY_DEVICE_TO_HOST;
+    } else if (SrcPtrType == sycl::usm::alloc::host ||
+               SrcPtrType == sycl::usm::alloc::unknown) {
+      assert(false && "Cannot copy image from host to host");
+    } else {
+      assert(false && "Unknown copy source location");
+    }
+  } else {
+    assert(false && "Unknown copy destination location");
```

Comment:
Does anything stop the users from hitting these asserts? In general, we only want to use asserts for cases that shouldn't be possible for the users to hit, as not only do asserts abort they are also disabled in production mode. Instead, we should be throwing exceptions if users try to do something they're not supposed to.

Reply:
Updated to throw exceptions.